### DISCUSSION TITLE
Dose SOCI support Oracle ODBC?

### DISCUSSION THREAD

**Discussion by 330305020** [SOCI/soci#1060] [General]
Hello gentlemen, I have a question. Dose SOCI support Oracle ODBC? And how? It's not mentioned in the document. Thx.

**Comment by zann1x** [comment]
The ODBC backend is supposed to work with any database that implements an ODBC driver. This includes databases that are not explicitly mentioned anywhere in the SOCI documentation. Oracle has an ODBC driver available, so I assume it will work.

Working with the database should then mostly look the same across all databases. Link `soci_core` and `soci_odbc` to your application (https://soci.sourceforge.net/doc/release/4.0/installation/) and connect via the ODBC backend (https://soci.sourceforge.net/doc/release/4.0/backends/odbc/#connecting-to-the-database).

**Reply by 330305020** [reply]
Thanks. I tried and it works. For Oracle ODBC (driver OracleClient21Home1_32bit), the connection string should be like "DSN=xx;UID=aa;PWD=bb;", and only "DSN=xx" is not working since when debugging VS projects, it will ask you to input password. For PostgreSQL ODBC (driver PostgreSQL Unicode), connection string "DSN=YY" is OK.

**Reply by zann1x** [reply]
The password prompt shouldn't pop up if all the necessary information is supplied. It sounds like you're missing something in the DSN for Oracle. But if you're sure, the prompt can be disabled via the option `odbc_option_driver_complete` (https://soci.sourceforge.net/doc/release/4.0/backends/odbc/#configuration-options).

**Reply by 330305020** [reply]
Thanks for your information. I tried the code as below in VS2017, the intension is to create session in pool:
    for (std::size_t i = 0; i < g__pool_size; ++i)
    {
      soci::session& sql = m_pSociConnectionPool->at(i);
      soci::connection_parameters parameters("odbc", "DSN=xx");
      parameters.set_option(soci::odbc_option_driver_complete, "0" /* SQL_DRIVER_NOPROMPT */);
      sql.open(parameters); 
    }

Then I have below exception: password is not provided
![image](https://github.com/SOCI/soci/assets/44339104/43165138-86c4-4baa-bff9-380dc3faf002)

I use below ODBC: OracleClient21Home1_32bit
![image](https://github.com/SOCI/soci/assets/44339104/cfeab7c2-986a-4617-ab19-81732fa9135e)

Each time you tried to "test connection" button in below window, you are asked to input password. i think That's the reason in my application I have exception if connection string is only "DSN=xx". Do you have some advice?
![image](https://github.com/SOCI/soci/assets/44339104/8c5af85c-ab83-4db4-968e-28345d607eb7)
![image](https://github.com/SOCI/soci/assets/44339104/39394869-9969-44d5-abdd-0f5f570b551d)

**Reply by zann1x** [reply]
> Each time you tried to "test connection" button in below window, you are asked to input password.

Your DSN is clearly missing the password then. You should either save the password in your DSN or provide it in the connection string as you've mentioned earlier.

I unfortunately don't use ODBC myself, so the official documentation of ODBC (and Oracle ODBC) should be able to provide you with more details about the necessary fields that a connection needs and how your database should be configured for that.

**Reply by 330305020** [reply]
Thanks.
